### ISSUE TITLE
Removal of SymEngine from `sympy.physics.mechanics`/`sympy.physics.vector`

### ISSUE THREAD

**Issue by brocksam** [sympy/sympy#25401]
This issue stems from a discussion on PR #25400.

Due to the increasingly drifting parity between SymPy and SymEngine, and inability for SymPy developers to actively fix all relevant SymEngine bugs (due to e.g. lack of C++ experience, delays due to release cycle etc.), we are proposing to unfortunately drop support for SymEngine within `sympy.physics.mechanics` and `sympy.physics.vector`.

@moorepants has suggested that we we might be able to do this "[without deprecation because it does not change results or API, it only has to do with performance](https://github.com/sympy/sympy/pull/25400#issuecomment-1643952943)."

We are making an assumption here that there are very few, if any, users of SymEngine with `sympy.physics.mechanics`/`sympy.physics.vector`. If you _do_ use SymEngine in this way, please let us know!

**Comment by asmeurer** [comment]
It would be worth writing to the mailing list about this.

**Comment by certik** [comment]
I was just pointed to this issue. I think it's a mistake to remove SymEngine, since it will drop the performance. However, since you and @moorepants are the main maintainers of the PyDy package, and if I understand your comments correctly, you do not use the SymEngine backend (because performance is less important than other priorities), and is slowing your development down, I can see how it makes sense to remove it from your perspective.

Let me ask the following questions and notes

* Last time I worked on this, the performance boost from SymEngine was huge, and it allowed to run some of the bigger examples like bicycle much quicker. Is that still the case?
* Is performance important to some users of this package?
* I noticed one of the issue was using `Symbol('r', positive=True)`. This API itself is not a problem by itself, but the current implementation attaches the assumptions to symbols and evaluates them as part of the core simplifications and we have not figured out how to make it fast. Instead, the assumptions must be implemented outside of the core, for things like "refine(e, Assumptions(...))", and I think there is probably a way to even keep the API of `Symbol('r', positive=True)`, but the core should not be doing it. That will actually converge the sympy and symengine way of doing it. This has been a long standing issue with SymPy, that unfortunately is still not resolved.
* Besides assumptions, what other issues with symengine/sympy compatibility did you encounter when developing this?
* Did you report bugs to SymEngine?
* In your experience, what path forward would you recommend, so that SymPy can use SymEngine as a backend, and it is easy to use and develop?

**Comment by oscarbenjamin** [comment]
> * Did you report bugs to SymEngine?

I have reported one or two and I think that they have been fixed.

The problem in general though is just that SymEngine and the parts of SymPy that it can replace when used as a backend will never be completely equivalent. Writing a single piece of code to work with SymPy expressions on the one hand or SymEngine expressions on the other is possible but there are just a lot of constraints involved in making it work and outside of relatively simple operations there will often be SymPy features that are needed that are not compatible between the two.

Essentially it is not well defined what exactly is the subset of SymPy functionality for which SymEngine could be used safely as an alternate backend (without anything breaking) and it is not clear which parts of SymPy could be constrained to work only with that subset of functionality.

**Comment by certik** [comment]
> Essentially it is not well defined what exactly is the subset of SymPy functionality for which SymEngine could be used safely as an alternate backend (without anything breaking) and it is not clear which parts of SymPy could be constrained to work only with that subset of functionality.

Yes, I think this is the main problem. @sylee957 and I discussed this at length, and we arrived to the conclusion that the only way to fix this is to create a "Symbolic API", a common API, that well defines what is the subset of sympy (and symengine and any other library that can provide a kernel) that you can use and things will work. It has to be enforced, ideally by construction, or at least with some tool, so that when you use the API, you will know that things will work with both backends. Similarly how the Array API works for NumPy.

**Comment by oscarbenjamin** [comment]
> Similarly how the Array API works for NumPy.

I have a hard time imagining what this would look like or how it could solve the problems here.

The array API is a bit different because there are usually clearly defined semantics for every operation even if different libraries made difference choices for which semantics to use. Almost nothing is clearly defined under the behaviour of SymPy's automatic evaluation: any one expression can just transform into something completely different. Likewise there are things like simplify etc that could never be well defined.

**Comment by certik** [comment]
Yes, you should not use simplify in library code. That's a well recognized lesson. There are differences like `2*(x+y)` vs `2*x + 2*y`, but otherwise things are actually very similar.

**Comment by moorepants** [comment]
> Last time I worked on this, the performance boost from SymEngine was huge, and it allowed to run some of the bigger examples like bicycle much quicker. Is that still the case?

I don't know what the current performance boost from SymEngine is for the large examples, but we are solving larger problems than the original bicycle example now with pure SymPy. Speeding up `.diff()` would still be helpful in SymPy.

> Is performance important to some users of this package?

SymPy Mechanics has this general workflow:

1. iterate on building a symbolic model until the model is correct
2. once the model is correct, code gen to fast numerics and move on

It can be annoying to have the iteration cycle in 1. take minutes instead of seconds, but it isn't something that isn't tolerable.

> I noticed one of the issue was using Symbol('r', positive=True). This API itself is not a problem by itself, but the current implementation attaches the assumptions to symbols and evaluates them as part of the core simplifications and we have not figured out how to make it fast. Instead, the assumptions must be implemented outside of the core, for things like "refine(e, Assumptions(...))", and I think there is probably a way to even keep the API of Symbol('r', positive=True), but the core should not be doing it. That will actually converge the sympy and symengine way of doing it. This has been a long standing issue with SymPy, that unfortunately is still not resolved.

This highlights the main issue, which is if SymEngine is to be a swap-able backend for SymPy, then every old and new feature of SymPy has to be added to SymEngine. For this to work, we'd really need a requirement in SymPy's PRs to also add equivalence to SymEngine before merging. But I don't think anyone would support something like that. I sure wouldn't.

> Besides assumptions, what other issues with symengine/sympy compatibility did you encounter when developing this?

There have been many.

> Did you report bugs to SymEngine

Yes, most.

> In your experience, what path forward would you recommend, so that SymPy can use SymEngine as a backend, and it is easy to use and develop?

If SymPy developers could add new features and fixes to SymPy and an equivalence would be **automatically** generated for SymEngine with the simultaneous PRs merged at the same time and SymeEngine's release cycle tied to SymPy's then it could work. Otherwise, SymPy development that uses SymEngine backend is always hindered by SymEngine keeping up (with functionality and as a proper released package).

> This library https://github.com/symforce-org/symforce uses both sympy and symengine, so performance is definitely important to them. They don't use pydy directly (except in one example), but if performance is important to them, why is it not to your users? I want to understand it better.

There are 3 performance needs for the SymPy mechanics based work:

1. When building a symbolic mathematical model you have to iterate interactively to construct you model. This is akin to the write+compile cycle for developing in compiled languages. If you have to wait too long for something to compile you get distracted and productivity is lower. This is the same for sympy mechanics models. The slowdown in this cycle is mostly due to the speed of `.diff()`.

2. We need `expr.diff()` to be faster because some of our symbolic expressions are so large that `.diff()` may not complete for hours/days on a modern laptop.

3. We generate numeric codes from extremely large symbolic equations and we need those codes to execute as fast as possible.

Point 1. is not a critical issue. It may be annoying that the iteration cycle is a bit long, but it isn't a deal breaker. Once the model is done you never have to do that again. This model building process usually is a few to hundreds of man-hour process depending on the model you are constructing and the iteration cycle cost is mostly irrelevant. The SymEngine backend can help here, but I've never bothered turning it on because it isn't so slow.

For Point 2, this can be a deal breaker for us and SymEngine could help here. But you can `cse()` an expression in SymPy and calculate the derivative of the cse'd version orders of magnitudes faster than SymEngine using only Python. This turns out to be the best solution because it is written in Python and becomes fast enough.

For Point 3, symengine doesn't really offer anything better than what sympy can do.

Over the last 8 years or so, pure SymPy has gotten faster such that the SymEngine speedup isn't worth the maintenance burden for what we do in the mechanics package.

**Comment by moorepants** [comment]
> This library https://github.com/symforce-org/symforce uses both sympy and symengine, so performance is definitely important to them. They don't use pydy directly (except in one example), but if performance is important to them, why is it not to your users? I want to understand it better.

It seems they have taken code from the PyDy repository and not honored PyDy's license.

**Comment by certik** [comment]
> It seems they have taken code from the PyDy repository and not honored PyDy's license.

Can you open up an issue in their repository if they took any code from PyDy? I am sure they'll fix it.

**Comment by certik** [comment]
@moorepants  Thanks for the comment above. 

> We need expr.diff() to be faster because some of our symbolic expressions are so large that .diff() may not complete for hours/days on a modern laptop.

This by itself seems sufficient enough to me to need SymEngine.

The question is how to do it. Above you suggested one way:

> If SymPy developers could add new features and fixes to SymPy and an equivalence would be automatically generated for SymEngine with the simultaneous PRs merged at the same time and SymeEngine's release cycle tied to SymPy's then it could work. Otherwise, SymPy development that uses SymEngine backend is always hindered by SymEngine keeping up (with functionality and as a proper released package).

But in a paragraph right above that you rejected it:

> For this to work, we'd really need a requirement in SymPy's PRs to also add equivalence to SymEngine before merging. But I don't think anyone would support something like that. I sure wouldn't.

I think this is too heavyweight as well. We need another solution.

----

I think the solution is a Symbolic API. I brainstormed with @sylee957 a little bit. I am inviting anybody to join the discussion. @moorepants would you have time to meet me over zoom? I would like to understand the issues you are facing. Roughly speaking, my goal is that fixes are done just once, in sympy, and one could swap the core and things will just work. I think it is possible, ideally with automatic checking, e.g., via LPython.

----

In the meantime, how feasible is to remove SymEngine, things will diverge, so it will not be possible to use SymEngine anymore, things will break. But then for the step 2., you convert the expression from sympy to symengine and then do the differentiation there?

**Comment by moorepants** [comment]
> But in a paragraph right above that you rejected it:

Those were two different suggestions. I would reject that we require all new SymPy changes force the PR author to also make a PR to SymEngine.

I wouldn't reject automating that (so a SymPy developer needs to know nothing about SymeEngine), but I don't think that solution is possible.

> But then for the step 2., you convert the expression from sympy to symengine and then do the differentiation there?

This could work for some cases, but does not work for all the internal calls to .diff() in the sympy mechanics classes.

**Comment by certik** [comment]
Ok. If you need speed inside the package, not just the final expression, then it seems to me you should keep symengine working. Do you or @brocksam have time to meet with me so that I can understand the issues that you are facing? I want to understand exactly what the compatibility problems are, so that we can ensure they are fixed (one way or the other). I was not aware that you had such big issues with SymEngine until literally 4 days ago. I think these issues can be fixed, now when I see that there is a problem. For example this would have been a great GSoC project to improve the compatibility. We have funding, we can organize things. That seems like a better solution than dropping the support. I agree that you should not spend weeks fixing this yourself, and that it should be mostly painless, so that you can focus on developing PyDy and not fight these differences.

**Comment by moorepants** [comment]
Fixing compatibility is a never ending chore that we don't have bandwidth or expertise for. Most PRs to change things in sympy.physics.vector/mechanics/biomechanics require us to waste time trying to get symengine tests to pass. Given that no one (not even us) likley uses the symengine backend, I don't see a reason to keep it around.

**Comment by moorepants** [comment]
I'm working on a problem for a conference paper that takes about 5 minutes to differentiate with SymPy, so I tried `USE_SYMENGINE=1` and get this:

```
In [1]: run linear_analysis.py
Defining reference frames.
Defining time varying symbols.
Orienting frames.
Defining constants.
Defining position vectors.
Defining holonomic constraints.
The holonomic constraint is a function of these dynamic variables:
[]
Defining kinematical differential equations.
Defining angular velocities.
Defining linear velocities.
Defining nonholonomic constraints.
Defining inertia.
Defining the rigid bodies.
Defining the forces and torques.
The nonholonomic constraints are a function of these dynamic variables:
[]
Generating Kane's equations.
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
File ~/Manuscripts/bmd2023-stable-low-speed-cargobike/src/linear_analysis.py:5
      1 import numpy as np
      3 from generate_numerical_funcs import get_sym_data
----> 5 sym_data = get_sym_data()
      7 v = 1.0  # m/s
      8 u6 = -v/sym_data.parameter_vals[-1]

File ~/Manuscripts/bmd2023-stable-low-speed-cargobike/src/generate_numerical_funcs.py:15, in get_sym_data()
     12 elif (os.path.getmtime('generate_eom.py') >
     13       os.path.getmtime('symbolics.pkl')):
     14     logging.info("Equation of motion file changed, recalculating.")
---> 15     from generate_eom import sym_data
     16 else:
     17     logging.info("Loading equations of motion from cache.")

File ~/Manuscripts/bmd2023-stable-low-speed-cargobike/src/generate_eom.py:396
    390 ###############
    391 # Kane's Method
    392 ###############
    394 print("Generating Kane's equations.")
--> 396 kane = mec.KanesMethod(
    397     N,
    398     q_ind,
    399     u_ind,
    400     kd_eqs=kinematical,
    401     q_dependent=q_dep,
    402     configuration_constraints=holonomic,
    403     u_dependent=u_dep,
    404     velocity_constraints=nonholonomic,
    405     constraint_solver='CRAMER',
    406 )
    408 Fr, Frs = kane.kanes_equations(bodies, loads=loads)
    410 r = sm.Matrix([T4, T6, T7, T9, T10])

File ~/src/sympy/sympy/physics/mechanics/kane.py:235, in KanesMethod.__init__(self, frame, q_ind, u_ind, kd_eqs, q_dependent, configuration_constraints, u_dependent, velocity_constraints, acceleration_constraints, u_auxiliary, bodies, forcelist, explicit_kinematics, kd_eqs_solver, constraint_solver)
    233 _validate_coordinates(self.q, self.u)
    234 self._initialize_kindiffeq_matrices(kd_eqs, kd_eqs_solver)
--> 235 self._initialize_constraint_matrices(
    236     configuration_constraints, velocity_constraints,
    237     acceleration_constraints, constraint_solver)

File ~/src/sympy/sympy/physics/mechanics/kane.py:326, in KanesMethod._initialize_constraint_matrices(self, config, vel, acc, linear_solver)
    324     B_ind = self._k_nh[:, :p]
    325     B_dep = self._k_nh[:, p:o]
--> 326     self._Ars = -linear_solver(B_dep, B_ind)
    327 else:
    328     self._f_nh = Matrix()

File ~/src/sympy/sympy/physics/mechanics/functions.py:713, in _parse_linear_solver.<locals>.<lambda>(A, b)
    711 if callable(linear_solver):
    712     return linear_solver
--> 713 return lambda A, b: Matrix.solve(A, b, method=linear_solver)

File symengine_wrapper.pyx:3860, in symengine.lib.symengine_wrapper.DenseMatrixBase.solve()

Exception: Unsupported method.
> /home/moorepants/Manuscripts/bmd2023-stable-low-speed-cargobike/src/symengine_wrapper.pyx(3860)symengine.lib.symengine_wrapper.DenseMatrixBase.solve()
```

Which is a typical issue. The solution is to add this method to Symengine (which is likely non-trivial to do), release a new version of Symengine, then make a new SymPy release that pins to this new Symengine. That process would mostly like have a year long cycle, which is not feasible even if I knew anything about updating Symengine.

**Comment by certik** [comment]
I think you should not be fixing this in SymEngine yourself, however you must be willing to report bugs.

The way it has to work is that you report a bug, and we fix it. I reported it here: https://github.com/symengine/symengine.py/issues/450, so don't worry about this one.

Can you go over the bugs that you reported here: https://github.com/symengine/symengine.py/issues and mark them with PyDy? I just created this label. 

--------

Let's discuss a more higher level design approach for this: I think we need to provide a Symbolic API and a way to check it. This Symbolic API will be supported by both sympy and symengine. You should not need to test with both sympy and symengine all the time. Rather, you should test with one, and if it works, it should just work with the other.

As an example above, this Symbolic API should not allow "method=linear_solver" if it is not supported by both sympy and symengine.

I agree that the current compatibility design is fragile. Let's fix it.

**Comment by moorepants** [comment]
> It will not take a year to fix this.

It will take a year to fix the bug in symengine, adopt the fix in sympy, release a new version of symengine, and then release a new version of sympy. This is what it takes before we have working code in downstream projects.

**Comment by moorepants** [comment]
> I think you should not be fixing this in SymEngine yourself, however you must be willing to report bugs.

When a SymPy PR fails because it isn't passing the symengine tests, that is exactly what a sympy dev is forced to do. The only two ways to move forward in a sympy pr are to 1. fix symengine and make a new symengine release or 2. disable symengine in the sympy tests. We do the later now.

**Comment by certik** [comment]
Right. I proposed a solution for this in my comment above.

**Comment by sylee957** [comment]
I think that it is a problem that SymPy developers come out with silent patch to add `CRAMER` to `solve`, without approval from symengine developers, pretend that the method exists there, and find that the code not works.

I generally disagree that we should add lots of options, or emphasis about options to the public facing API, 
for example, nullary functions like `solve()` or `rref()` should be useful and sufficient, 
other options should really be 'optional', than being major part of the way how to use the function.
or should be discussed in more formal settings if we really need it or should be implemented in SymEngine.

I think that most of the options were introduced to make users work around the performance issues in SymPy, 
however, the performance potential of SymEngine can be much more higher such that it can avoid that indirection.

**Comment by moorepants** [comment]
> I think that it is a problem that SymPy developers come out with silent patch to add CRAMER to solve, without approval from symengine developers, pretend that the method exists there, and find that the code not works.

There are no silent patches to SymPy, they are all public as PRs in this repository. If you are suggesting that all SymPy PRs should get approval form Symengine developers, that seems quite extreme and would slow down SymPy development immensely.

**Comment by moorepants** [comment]
As a reference, I have not been able to find any evidence on Github that anyone uses symengine with the mechanics package:

https://github.com/search?q=use_symengine+mechanics&type=code&p=1

**Comment by oscarbenjamin** [comment]
> I generally disagree that we should add lots of options, or emphasis about options to the public facing API,
> for example, nullary functions like `solve()` or `rref()` should be useful and sufficient,

I do agree that it would be better to have fewer options and that it would be better if the interface was more minimal. Many things that are methods defined on classes for example would be better as separate functions so that the interface exposed by e.g. Expr or Matrix would be more minimal. I will be interested to see what a "Symbolic API" looks like and I also think that if we are potentially talking about a SymPy 2.0 then there can be an opportunity to change much of the API and to make it simpler and more well defined. If designing a symbolic API means that the API ends up having a better design then that is a worthwhile exercise regardless of compatibility between SymPy and SymEngine.

We need to be clear though that the "API" that SymPy and SymEngine have has a huge surface and so even with a big reduction in complexity it will never be possible for both SymPy and SymEngine to be completely compatible at the API level that `USE_SYMENGINE` switches out. We also need to be clear that unlike the Array API which will probably be implemented by say 10 different libraries there will never be a third implementation of the Symbolic API unless it is *drastically* different from the API that SymPy and SymEngine currently expose for top-level use by users. If I was attempting to create an alternative to SymPy and SymEngine I would not even attempt to match their APIs at all regardless of whether there was a document that tries to formalise that API.

Regardless of any symbolic API the `USE_SYMENGINE` mechanism cannot be made to work because it is swapping out too much in one go with no compatibility layer that can smooth out the differences. The only way that it could ever work is if SymPy's pure Python core was dumped and SymEngine became a hard dependency. If that is what anyone proposes to do then they need to take a good look at how much is currently missing from SymEngine that would be needed to make it possible.

If it is not possible to extend SymEngine from Python to add e.g. new expression types then SymEngine would need to have *all* of the expression types that SymPy has which means that it is missing hundreds of classes:
```python
In [22]: class cos(se.Expr):
    ...:     pass
    ...: 

In [23]: cos(1)
Out[23]: Segmentation fault (core dumped)
```
How is downstream code that depends on SymPy supposed to add new expression types that are not implemented in SymPy or SymEngine? If SymEngine has no way to do that then it cannot replace SymPy's core.

Also if the proposal is for SymPy to adopt SymEngine and drop its pure Python core then I would demand significant changes to SymEngine to support unevaluated expressions. Otherwise in its current form SymEngine makes unevaluated expressions impossible and would prevent SymPy from solving many of its existing problems with them.

In gh-25495 I made it possible to swap out the internal implemention of dense DomainMatrix so that we can use Flint matrices when the domain is ZZ or QQ (integers or rational numbers). What that means is that without flint you have:
```python
In [4]: dM = Matrix([[1, 2], [3, 4]]).to_DM().to_dense()

In [5]: dM
Out[5]: DomainMatrix([[1, 2], [3, 4]], (2, 2), ZZ)

In [6]: dM.rep
Out[6]: DDM([[mpz(1), mpz(2)], [mpz(3), mpz(4)]], (2, 2), ZZ)
```
Here `DomainMatrix` has an internal `DDM` instance which is the pure Python dense representation of this matrix over ZZ. With flint (`pip install python-flint` and set `SYMPY_GROUND_TYPES=flint`) you have:
```python
In [1]: dM = Matrix([[1, 2], [3, 4]]).to_DM().to_dense()

In [2]: dM
Out[2]: DomainMatrix([[1, 2], [3, 4]], (2, 2), ZZ)

In [3]: dM.rep
Out[3]: DFM([[1, 2], [3, 4]], (2, 2), ZZ)

In [4]: dM.rep.rep
Out[4]: 
[1, 2]
[3, 4]

In [5]: type(dM.rep.rep)
Out[5]: flint.types.fmpz_mat.fmpz_mat
```
So here the internal representation is python-flint's `fmpz_mat`.

The point to notice here is the class `DFM` that sits in between `DomainMatrix` and `fmpz_mat`. It provides the same interface as `DDM` but implements that interface using an internal `fmpz_mat` or `fmpq_mat` depending on whether the domain is `ZZ` or `QQ`.

Now it could be possible to make it so that both `DDM` and `fmpz_mat` have the exact same interface so that we would not need the `DFM` class as a wrapper around the `fmpz_mat`. However it is *much* easier to make the two things compatible if there is a compatibility layer between the interface that `python-flint` exposes and the interface that `DomainMatrix` expects. There are all sorts of reasons that this compatibility layer is useful:

- The compatibility layer can account for different versions of python-flint that might change its interface (this immediately solves the problem that Jason refers to about needing to make a change in SymEngine, wait for a new SymEngine release etc).
- The code within the rest of SymPy can be much simpler if nothing else ever needs to know about the difference between `DFM` and `DDM`.
- We can easily change the interface of both `DFM` and `DDM` together without needing to coordinate in any way with python-flint.

Over time we might want to simplify the `DFM` compatibility layer by making python-flint's interface more consistent with the interface that `DomainMatrix` expects either by changing `DDM` and `DFM` or by changing `fmpz_mat`. I would never propose though to remove `DFM` entirely and just use `fmpz_mat` directly just because it will always be helpful to have that layer that can be used to maintain compatibility.

Now let's contrast this situation with the layer that `USE_SYMENGINE` tries to swap out:

The case of matrices of simple integers/rationals and very well defined operations over those matrices like `inv`, `det`, etc is extremely simple in comparison with things like `Expr/Matrix`. There is never a need to worry about differences like `2*x + 2*y` vs `2*(x + y)` etc because we will always just represent everything as simple integers. By contrast even the simplest operations involving Expr can be implemented in different ways:
```python
In [18]: import sympy as sym

In [19]: import symengine as se

In [20]: print(se.sin(se.pi/10))
sin((1/10)*pi)

In [21]: print(sym.sin(sym.pi/10))
-1/4 + sqrt(5)/4

In [3]: se.sympify('2*(x + y)')
Out[3]: 2*(x + y)

In [4]: sym.sympify('2*(x + y)')
Out[4]: 2⋅x + 2⋅y
```
For the API at this level to be *completely* compatible SymEngine and SymPy would need to have completely equivalent evaluation rules. No one can even tell you or define what those evaluation rules currently are for SymPy because they are distributed over hundreds of classes and hundreds of thousands of lines of code. It would be nice if these things were more well defined or if these evaluation rules could be documented etc. However as long as all of these evaluation rules are hard-coded across hundreds of `__new__` methods separately in both SymPy and SymEngine differences like the cases above will always exist.

The interface between `DomainMatrix` and `DDM` was deliberately designed from the outset with the intention that it should be possible to swap out different implementations of that interface and so as much as possible the methods exposed in the interface are kept simple. For example `Matrix.__getitem__` is extremely complicated and so I did not implement `DDM.__getitem__` at all. Instead `DDM` provides `getitem`, `extract`, `extract_slice`, `setitem` etc. Each of these methods is extremely simple and does not need to worry about all of the complexity that a `__getitem__` method would need to provide. By contrast Expr/Matrix were not designed in a way that it would ever be possible to make multiple compatible implementations of the same interface. Practically none of the Expr/Matrix methods has a simple interface or unambiguous semantics.

Even in the very simple case of DomainMatrix and python-flint's matrices I added the `DFM` compatibility layer. That makes it possible for me to control any differences between python-flint and SymPy without any other SymPy contributors needing to even know about Flint or the `DFM` class: there are no incompatibilities so there is no need to worry about them. Further changes analogous to this could improve the performance of e.g. SymPy's mechanics module without causing any difficulties for the people who work with or improve the mechanics module.

Does the `DFM` layer bring a little overhead and cause things to be a little slower than if `fmpz_mat` was used directly by `DomainMatrix`?

Yes, the layer adds some overhead but that overhead can be made trivial in the context of everything else.

The `USE_SYMENGINE` mechanism has no compatibility layer. It just tries to swap out a huge API surface that was never designed to be swapped. Also it is not possible to implement a compatibility layer over the top of the interface that SymEngine provides and the reason is because of automatic evaluation: transforming e.g. `sin(pi/10)` into `-1/4 + sqrt(5)/4` is irreversible. A compatibility layer over the top of SymEngine would need to be able to control what the expressions are doing but there is no way to control the automatic evaluation.

If SymEngine provided unevaluated expressions then it would be at least possible to implement a compatibility layer. In that situation though the compatibility layer would not use *any* of SymEngine's existing API: it would use the new unevaluated expression API and the evaluation rules would be controlled from SymPy.

**Comment by certik** [comment]
Indeed, the problem is that currently we swap a huge API (SymPy) for another huge API layer (SymEngine), and then it fails when you provide a different option to solve(). I agree with @sylee957.

> If you are suggesting that all SymPy PRs should get approval from Symengine developers, that seems quite extreme and would slow down SymPy development immensely.

No, that is not what he is suggesting. 

As I suggested in https://github.com/sympy/sympy/issues/25401#issuecomment-1722143339, the way the Symbolic API has to work is that there must be a layer that defines what is allowed and is tested with tests. We then trivially ensure both SymPy core, SymEngine (and other other library that we decide to write) passes it.

Then when you build using this API, you are by construction guaranteed your code will work with both SymPy and SymEngine (and any other core).

Regarding Matrix.solve(), we have two ways to solve it:

* either we decide it is part of this API, in which case we ensure it works with all kernels and then you can use it. If you provide an option to the function that is not supported, then you get an error. You will have two options: either you don't use the option you wanted, or you use SymPy directly. If you use SymPy directly, then you lose the guarantee that it works with SymEngine. So if you want compatibility, you must use the API.

* Or we decide this should be built on top of the API. In that case we implement this in sympy on top. And then by construction it is guaranteed to work.


The above approach allows us to gradually port sympy on top of the API, step by step, without ever breaking things. You will always be able to use SymPy directly via the old API, as well as the Symbolic API.

Regarding who else will implement this? We already have 3 implementations:

* current SymPy
* SymEngine
* LPython symbolics

There will likely be many more, including protosym, various old projects like sympycore, etc. All of them use almost the same API in Python.

This API would also be interesting for Sagemath and really any symbolics in Python.

I agree I also prefer smaller classes and global functions.

Yes, SymEngine of course allows user defined classes defined in Python (or any other language). We can design this as part of the Symbolic API how to standardize it.

Regarding unevaluated expressions, all we need is to design how it should work with the API, how you would call it as a user. For PyDy I think it is not needed.

Yes, one could think of it as SymPy 2.0. As long as SymPy 1.0 and 2.0 works side by side (via adapters) and we don't break things, then this can work.

Regarding standardizing what simplifications the API should do, we need to discuss it. Again, we have two options:

* Either we standardize exactly what gets done. I think what SymEngine is doing is better, it does less simplifications by default. We could go all the way to what protosym does and do no simplifications by default, and if we can show with actual code that we can keep performance on all relevant benchmarks, we can do that. If we can't keep performance, we have to do some simplifications by default.
 
* Or we allow the expressions to vary across various cores. In that case we specify what is allowed, such as both `2*(x + y)` and `2*x + 2*y`. And if you as a use want a specific form, you have to call some function, say "expand()", then it woul be guaranteed to return `2*x + 2*y`.


The hard part is to figure out all the details, but this can be done if there is will from all parties to design this API that works for all of us.

**Comment by oscarbenjamin** [comment]
> Regarding standardizing what simplifications the API should do, we need to discuss it. Again, we have two options:

There is a third option:

- The standardised API does not define any simplification at all and uses only unevaluated expressions. The API only defines operations that manipulate those expressions. Any evaluation rules would be defined on top of this API but the API provides the mechanisms to implement those rules.

If you want to define a swappable interface like the one between DomainMatrix and DDM then you need that interface to be as simple as possible. Any interface that includes anything like SymPy and SymEngine's current automatic evaluation as an implicit part of every operation will always be too complex to standardise.

What this would mean though is that the standardised "Symbolic API" would *not* be an API for most end users. It would be the equivalent of `DFM` which is two levels down from the `Matrix` interface that users use. Functions like `rref`, `solve` etc could be standardised only if they are defined at the computational algebra level e.g. like the polys and DomainMatrix vs python-flint. An `rref` for arbitrary symbolic expressions is probably already too high-level to be something that could be standardised.

Note that to say that the API does not define simplification does not mean that the API would not provide a way to have automatic simplification. The rules for simplification would have to be something that can be passed into the API like:
```python
rules = {
   Pow(a, 1): a,
   Pow(a, 0): One,
   Mul(a, a): Pow(a, 2),
   ...
}
def evaluate(expr):
    return apply_rules(expr, rules)
```
The rules have to be something that is explicitly controllable while using the API rather than something that is implicit in every operation exposed by the API.

Note that there are many other reasons why it would be better to have the rules be defined like this regardless of speed or SymPy vs SymEngine compatibility. If the rules and the assumptions associated with the rules were defined like this then it would be possible to use the new assumptions in combination with the evaluation rules. Also having declarative rules makes the whole system much more easily extensible (users can define their own expression types and rules) and much less bug-prone.

**Comment by oscarbenjamin** [comment]
> If the rules and the assumptions associated with the rules were defined like this then it would be possible to use the new assumptions in combination with the evaluation rules.

To clarify what I mean by this many of the rules would need to have conditional guards like:
```python
rules = {
    # 0/x -> 0 if x != 0
    If(Ne(x, 0), Div(0, x) >> 0),
    ...
}
```
Currently these things are implemented in e.g. `Mul.flatten` but it needs to use the assumptions system to resolve `x.is_zero`. With declarative rules as above you can do:
```python
expr = evaluate(expr, rules, assumptions={x > 0})
```

**Comment by certik** [comment]
Regarding your third approach in https://github.com/sympy/sympy/issues/25401#issuecomment-1722240386, yes, it's a valid approach. However, I would like to see a prototype of this API working with a few representative benchmarks, matching (or beating!) SymEngine's speed. So far I have not seen a system that can do that. Your protosym can beat SymEngine one benchmarks that don't require simplification, but so far it's much slower on benchmarks that require it (which I argue is the majority of use cases).

Yes, assumptions can't be in the core for performance. They must be applied on top using `refine(expr, assumptions={x > 0})`, which is similar to your idea.

**Comment by oscarbenjamin** [comment]
> However, I would like to see a prototype of this API working with a few representative benchmarks, matching (or beating!) SymEngine's speed.

Turning this around: I would like to see a demonstration of SymEngine's performance for expression types that are defined from Python. Actually I would just like to see a demonstration of how to add new expression types from Python when using SymEngine at all because SymPy will never be able to use SymEngine if it cannot provide this.

Looking in the SymEngine tests there are a couple of tests for subclassing Symbol but nothing else and when I try to subclass Function or Expr it crashes:
```pycon
In [1]: import symengine as se

In [2]: class tan(se.Function): pass

In [3]: tan(1)
Out[3]: Segmentation fault (core dumped)
```
Currently SymPy has 999 Basic subclasses and SymEngine has 83 by my count:
```python
In [18]: import symengine as se

In [19]: objs = [getattr(se, n) for n in dir(se)]

In [20]: len([o for o in objs if isinstance(o, type) and issubclass(o, se.Basic)])
Out[20]: 83
```
The set of expression types defined in SymPy is always going to be larger than the set defined in SymEngine. So for SymPy to use SymEngine in general there would need to be a way for SymPy to add expression types (e.g. new mathematical functions) that SymEngine does not have. Likewise many downstream libraries that depend on SymPy define new expression types and so that will always need to be supported.

It is already too difficult to add new expression types and we need to make that easier rather than harder. Ideally it would be so easy that end users could do it e.g. this is how you do it in Mathematica:
```python
In[8]:= f /: Derivative[1][f] = g
Out[8]= g

In[9]:= D[f[x], x]
Out[9]= g[x]

In[10]:= D[f[x]^2, x]
Out[10]= 2 f[x] g[x]
```
We just invented two new functions `f` and `g` and defined the differentiation rule for `f`. Now Mathematica can compute derivatives involving these functions.

The SymPy way is much too complicated to be something that someone would just do in an interactive session:
https://docs.sympy.org/latest/guides/custom-functions.html#differentiation

**Comment by moorepants** [comment]
> either we decide it is part of this API, in which case we ensure it works with all kernels and then you can use it. If you provide an option to the function that is not supported, then you get an error.

Not getting an error would be preferable. If the symengine backend worked in a way that it fell back to SymPy whenever it couldn't provide functionality, then there would be no maintenance burden.

> Regarding unevaluated expressions, all we need is to design how it should work with the API, how you would call it as a user. For PyDy I think it is not needed.

We would very much prefer this for the PyDy use cases, because it would not only speed up various operations but allow us easy rewriting for code generation on these unevaluated trees that could include very top-level mathematical operations.

**Comment by moorepants** [comment]
Maybe a different issue could be opened for this current discussion, or even a sympep. This issue is much more narrow. We desire to remove the maintenance burden of dealing with the symengine backend in physics.vector/mechanics, i.e. a short term solution that is tangent to the broader discussion on speeding up SymPy and SymPy 2.0. I have yet to find any evidence that people use the symengine backend in these packages and the benefit it offers currently does not outweigh the cost.

Many of the pull requests in this search had issues getting the symengine tests to pass and we often just disable them:

https://github.com/search?q=repo%3Asympy%2Fsympy+symengine+mechanics&type=pullrequests

These are many of the PRs we've made in the last year for the GSoC and CZI projects. It just costs us unnecessary development time.

I've opened #25617 to remove the backend. It can always be added back when it is in a state that is not a maintenance burden.

**Comment by certik** [comment]
@moorepants I discussed with @asmeurer in person. I think if you don't want SymEngine at the current state, then you can remove it. I am sad, since it took me and other a lot of work to get it working with PyDy. But I think the way it is currently used in SymPy makes it hard for you to test.

We will instead pick some smaller module and try to get it working with the Symbolic API. I was thinking limits. I would like to make it work with LPython, that would make limits as fast as if it was implemented in C++ and symengine. But it would be in Python, and still compatible with SymPy.

P.S. @oscarbenjamin we support both in SymEngine tests, you can play with it. I already sent you links for tests for sympy functions in private correspondence. And here is a link for the function "f(x)": https://github.com/symengine/symengine.py/blob/master/symengine/tests/test_functions.py.

**Comment by isuruf** [comment]
@oscarbenjamin, SymEngine type subclassing is not supported at the moment, but SymEngine can use arbitrary SymPy functions. For eg:

https://github.com/symengine/symengine.py/blob/8d5d63c7842ca92f7f3c5436f9825ace984e4af0/symengine/tests/test_functions.py#L147-L170

Here `Ynm` is a SymPy function that SymEngine does not have support natively, but SymEngine knows how to differentiate, substitute, eval, etc.

```python
In [1]: import sympy, symengine

In [2]: n, m, theta, phi = sympy.symbols("n, m, theta, phi")

In [3]: r = symengine.Integer(2) * sympy.Ynm(n, m, theta, phi)

In [4]: type(r)
Out[4]: symengine.lib.symengine_wrapper.Mul

In [6]: r.diff(theta)
Out[6]: 2*(m*cot(theta)*Ynm(n, m, theta, phi) + exp(-I*phi)*sqrt((-m + n)*(1 + m + n))*Ynm(n, 1 + m, theta, phi))

In [8]: r.subs({n: 1, m: 1, theta: symengine.pi/4, phi: symengine.pi/8})
Out[8]: 2*Ynm(1, 1, (1/4)*pi, (1/8)*pi)

In [9]: s = r.subs({n: 1, m: 1, theta: symengine.pi/4, phi: symengine.pi/8})

In [10]: s.evalf()
Out[10]: -0.45140986028071 - 0.186980086317214*I

In [11]: s._sympy_()
Out[11]: 2*Ynm(1, 1, pi/4, pi/8)

In [12]: s._sympy_().evalf()
Out[12]: -0.45140986028071 - 0.186980086317214*I

```

**Comment by moorepants** [comment]
> Not getting an error would be preferable. If the symengine backend worked in a way that it fell back to SymPy whenever it couldn't provide functionality, then there would be no maintenance burden.

Is this something that is possible without a lot of work?

**Comment by oscarbenjamin** [comment]
> Here `Ynm` is a SymPy function that SymEngine does not have support natively, but SymEngine knows how to differentiate, substitute, eval, etc.

Thanks @isuruf.

So if I understand correctly SymEngine has a compatibility layer that can embed Python SymPy expressions and then presumably just calls the Python methods internally. These expressions still need to be defined using the existing Python core though (which therefore could not be removed) and SymEngine cannot make operations with these kinds of expressions any faster than they would be without SymEngine.

Do you envisage any alternative way of extending SymEngine from Python in future?

**Comment by isuruf** [comment]
> SymEngine cannot make operations with these kinds of expressions any faster than they would be without SymEngine.

Not really. It's only the function that cannot be made faster, but the expression itself can become faster.

```python
In [1]: import sympy, symengine

In [3]: n, m, theta, phi = sympy.symbols("n, m, theta, phi")

In [4]: symengine_expr = symengine.Integer(2) * sympy.Ynm(n, m, theta, phi)

In [6]: sympy_expr = sympy.sympify(symengine_expr)

In [7]: symengine_theta = symengine.sympify(theta)

In [8]: %time a = symengine_expr.diff(symengine_theta, 5);
CPU times: user 59.9 ms, sys: 421 µs, total: 60.4 ms
Wall time: 59.8 ms

In [9]: %time a = sympy_expr.diff(theta, 5);
CPU times: user 499 ms, sys: 2.51 ms, total: 502 ms
Wall time: 501 ms
```

(SymEngine offloads the comptutation of the derivative of the Ynm function itself to SymPy and does the rest natively. If the arguments themselves are large there will be performance issues, but that can be optimized further as well.)


> Do you envisage any alternative way of extending SymEngine from Python in future?

Yes. We already have a way to extend Symbol class, and that can be extended to Function class as well.

**Comment by isuruf** [comment]
> Is this something that is possible without a lot of work?

@moorepants, this cannot be done easily as far as I know. For eg: if wanted to support `Matrix.solve(method="CRAMER")` we can add a fallback to SymPy for all unknown methods. However we would have to do this for each function.

**Comment by oscarbenjamin** [comment]
> If the arguments themselves are large there will be performance issues, but that can be optimized further as well.

It would be better if the methods exposed were different. The current `_eval_` methods for differentiation are poorly designed. Ideally SymEngine could just ask `Ynm` what its partial derivatives are in general form rather than asking the `Ynm` object to recursively differentiate its arguments.

**Comment by isuruf** [comment]
Yes, that was my thought as well. When having custom classes in SymEngine, I would prefer the `_eval_` methods to be restricted in what they can do compared to SymPy's design.

**Comment by oscarbenjamin** [comment]
The other problem is the lack of functions as objects. Ideally you would have something like `diff(sin) -> cos` but the fact that `sin` and `cos` are classes means that it is difficult to define symbolic rules with them without turning them into expressions like `sin(x).diff(x) -> cos(x)`.

In Wolfram language there are two differentiation operators `D` and `Derivative`:
```python
In[7]:= Derivative[1][Sin]
Out[7]= Cos[#1]&

In[8]:= Derivative[1][f]
Out[8]= f'

In[9]:= D[Sin[x], x]
Out[9]= Cos[x]

In[10]:= D[f[x], x]
Out[10]= f'[x]

In[11]:= D[f[x], x] // FullForm
Out[11]//FullForm= Derivative[1][f][x]
```
Here `f` can be a symbolic object separately from `f[x]`. That can't really work properly in SymPy's design because if `f(x)` is an instance of `Basic` then `f` needs to be a subclass of `Basic` and therefore `f` is not a proper symbolic expression on its own.

**Comment by certik** [comment]
Regarding treating "f" or even just "sin", "cos" as entities of their own, so that you can do things like "e = sin + cos + f" and then "e(x)" would become "sin(x) + cos(x) + f(x)", that would have to be built on top. It seems like a different class of expression. You don't want to have any performance overhead for the more common "sin(x)+cos(x)" case. The way I would do it is to implement a separate algebra / classes to handle "sin + cos", and then when applying it to "x", I would return the current SymPy/SymEngine "sin(x) + cos(x)".

**Comment by oscarbenjamin** [comment]
> Regarding treating "f" or even just "sin", "cos" as entities of their own, so that you can do things like "e = sin + cos + f"

Other people might want to do that but that is not what I want this for. the point is to be able to define a derivative without a symbol:
```mathematica
In[21]:= f'
Out[21]= f'

In[22]:= f' // FullForm
Out[22]//FullForm= Derivative[1][f]

In[23]:= f'[0]
Out[23]= f'[0]

In[24]:= f'[0] // FullForm
Out[24]//FullForm= Derivative[1][f][0]
```
SymPy has no way to represent `f'(0)` directly.

Differentiation rules could be much better expressed if a function could be differentiated without a symbol.

**Comment by certik** [comment]
It's the same thing in my mind. You define operators like "+" and "Derivative" to work on functions that are not "applied" to "x".

**Comment by oscarbenjamin** [comment]
The `+` case is just not a very important reason for wanting it whereas the derivative case is a significant limitation. Essentially higher-order symbolic operations are not expressible in SymPy and differentiation should naturally be expressed as a high-order symbolic operation: an operator on functions rather than expressions.

**Comment by sylee957** [comment]
I think that it is valid that SymEngine maintainers should be notified about this issue,
however, I don't think that we need to continue discussion about some technical details about derivatives or problems about the APIs here.

Also, the derivative extension APIs seems to be a good candidate for SymPEPs, 
becuase it may need nontrivial change to the SymPy core,
and the decisions about the API can affect the performance, 
so I don't expect it to be done in one-pass contribution in SymPy directly.

**Comment by oscarbenjamin** [comment]
Good point @sylee957 and @moorepants already said above that it would be good to have a separate issue for this. I don't want the more general discussion to stop but this issue was opened for something more limited (but not unimportant) that also needs to be resvoled.

I think that it is agreed from above that the `USE_SYMENGINE` mechanism can be difficult to use for SymPy developers who are trying to maintain compatibility with both APIs. The people working on `sympy.mechanics` have said that they find this to be overly constraining in their work and that they would prefer to remove that constraint. I think we should agree that they can do what they need to and then continue the discussion about SymPy/SymEngine compatibility elsewhere.

**Comment by certik** [comment]
> > Is this something that is possible without a lot of work?
> 
> @moorepants, this cannot be done easily as far as I know. For eg: if we wanted to support `Matrix.solve(method="CRAMER")` we can add a fallback to SymPy for all unknown methods. However we would have to do this for each function.

@moorepants I thought about this, and I think it can be done. Once we have a Symbolic API layer (details to be figured out), the Matrix would be implemented on top (most likely). So you would just call it. If you specify a "method" that the symbolic API supports, then it would be routed via it. If you specify a "method" that is not supported, then it would route it via SymPy (doing the appropriate conversions as needed).

So I think it might work. What we don't want to do is to implement it in `symengine.py`, as @isuruf said. We also don't want to silently get slow by implicitly routing via SymPy. I think the way the Symbolic API has to work is that either the operation is supported with all backends (and if SymEngine is used, it will always be fast), never any implicit conversions; or it is not supported (like `method="CRAMER"`) in which case it must give an exception (even with the SymPy backend). But building the implicit routing via SymPy can be built on top as I mentioned above. That I think could work.

To move forward, I am fine if you remove SymEngine from PyDy for now (but not SymPy), since as you said, you don't really use it, we don't know if there are people using it (I wasn't able to find GitHub projects either), and as you said, you don't want it (all things considered). 

And I'll continue working on making progress on the Symbolic API via LPython. The work is progressing very nicely, here is an example how it works: https://github.com/lcompilers/lpython/blob/0a5d89b53e856f4b5998464e0d4936d9328e4896/integration_tests/symbolics_05.py, this code compiles today with LPython+Symengine, as well as with regular CPython+SymPy. Our next step is to start porting some very small subset of SymPy, we are thinking of starting with limits. Say in a year or two, once this matures, the workflow will be that if you want speed, you just ensure that your Python code compiles with LPython. If it does, it is guaranteed to fully compile and run with SymEngine, as well as with CPython/SymPy.

**Comment by oscarbenjamin** [comment]
> I am fine if you remove SymEngine from PyDy for now (but not SymPy)

It is not completely clear what you mean by this. The PR gh-25617 is the change that this issue was originally opened to discuss. It does not remove the `sympy.core.backend` module but it does remove all of the places in the mechanics and vector modules that import from that backend.

**Comment by moorepants** [comment]
> @moorepants I thought about this, and I think it can be done. Once we have a Symbolic API layer (details to be figured out), the Matrix would be implemented on top (most likely). So you would just call it. If you specify a "method" that the symbolic API supports, then it would be routed via it. If you specify a "method" that is not supported, then it would route it via SymPy (doing the appropriate conversions as needed).

I can imagine it working for an optional kwarg for a single method, but I mean can it work more generally? The ideal situation is that SymPy developers don't have to know anything about Symengine and SymPy users never get a Symengine error if something is missing when they set `USE_SYMENGINE=1`. We would want Symengine to be used when it can and fall back gracefully to SymPy when it can't.

**Comment by certik** [comment]
> I can imagine it working for an optional kwarg for a single method, but I mean can it work more generally? The ideal situation is that SymPy developers don't have to know anything about Symengine and SymPy users never get a Symengine error if something is missing when they set USE_SYMENGINE=1. We would want Symengine to be used when it can and fall back gracefully to SymPy when it can't.

I think so. I think the design I suggested above would work exactly like this.

**Comment by oscarbenjamin** [comment]
We should take this discussion to new issues and merge the PR for mechanics.

I suggest opening a new issue with a concrete proposal for how SymPy and SymEngine can interact.

**Comment by moorepants** [comment]
Thanks everyone for the discussion and action.
